src/util/LimitedAllocator.h

```diff
@@ -0,0 +1,164 @@
+ * This class is threadsafe via ad_utilitty::Synchronized
+ */
+class AllocationState {
+ public:
+  AllocationState() = delete;
+  using T =
+      std::shared_ptr<ad_utility::Synchronized<AllocationLimits, SpinLock>>;
+  explicit AllocationState(T ptr) : ptr_{std::move(ptr)} {}
+  T& ptr() { return ptr_; }
+  const T& ptr() const { return ptr_; }
+
+  friend bool operator==(const AllocationState& a, const AllocationState& b) {
+    return a.ptr_ == b.ptr_;
+  }
+
+ private:
+  T ptr_;
+};
+}  // namespace detail
+
+// setup a shared Allocation state. For the usage see documentation of the
+// Limited Allocator class
+inline detail::AllocationState makeAllocationState(size_t n) {
+  return detail::AllocationState{std::make_shared<
+      ad_utility::Synchronized<detail::AllocationLimits, SpinLock>>(n)};
+}
+
+/**
+ * @brief An Allocator that limits the amount of memory totally allocated at the
+ * same time
+ *
+ * The actual allocation is done by a std::allocator<T> which is wrapped by
+ * checking the limit
+ *
+ * Memory allocated by copies of an Allocator will also count towards the limit
+ * To use it, construct a first allocator by calling
+ * LimitedAllocator{makeAllocationState(limit)} and then pass copies of this
+ * allocator to your containers. You can also Create Allocators for different
+ * types respecting the same total memory pool size, e.g.
+ *
+ * auto state = makeAllocationState(limitInBytes);
+ * auto allocInt = LimitedAllocator<int>{state};
+ * auto limitedIntVec = std::vector<int, LimitedAllocator<int>>{allocInt};
+ * auto allocString = LimitedAllocator<string>{state};
+ * auto limitedStringVec = std::vector<string,
+ * LimitedAllocator<string>>{allocString};
+ *
+ * // now the total amount of memory allocated by limitedIntVec and
+ * limitedStringVec may never exceed limitInBytes
+ *
+ * @tparam T the type of Elements that this allocator allocates
+ */
+template <typename T>
+class LimitedAllocator {
+ public:
+  using value_type = T;
+
+  // These special type aliases are necessary for using this allocator with
+  // the STL. We always want it to be propagated to the target of copy and move
+  // operations, s.T. the copy also counts towards the Limit.
+  using propagate_on_container_copy_assignment = std::true_type;
+  using propagate_on_container_move_assignment = std::true_type;
+  using propagate_on_container_swap = std::true_type;
+
+ private:
+  detail::AllocationState state_;  // shared number of free bytes
+  std::allocator<T> alloc_;
+
+ public:
+  /// obtain an AllocationState by calls to makeAllocationState()
+  explicit LimitedAllocator(detail::AllocationState s) : state_(std::move(s)) {}
+  LimitedAllocator() = delete;
+
+  // TODO<C++20> : the exact signature of allocate changes
+  // the required interface of an allocator
+  T* allocate(std::size_t n) {
+    // first check if we have enough space and reserve it (otherwise throw)
+    state_.ptr()->wlock()->allocate(n * sizeof(T));
```

Comment:
memoryLeft_.ptr()->wlock()->decrease_if_enough_left(...);

Reply:
My IDE fortunately supports consistent renaming of classes/functions/parameters etc.